gradoop-flink/src/main/java/org/gradoop/flink/algorithms/gelly/randomjump/functions/GetVisitedSourceTargetIdsFlatMap.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright © 2014 - 2018 Leipzig University (Database Research Group)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradoop.flink.algorithms.gelly.randomjump.functions;
+
+import org.apache.flink.api.common.functions.FlatMapFunction;
+import org.apache.flink.api.java.functions.FunctionAnnotation;
+import org.apache.flink.util.Collector;
+import org.gradoop.common.model.impl.id.GradoopId;
+import org.gradoop.common.model.impl.pojo.Edge;
+
+/**
+ * Retrieves the source- and target-id from an edge, if this edge was visited, determined by the
+ * boolean property.
+ */
+@FunctionAnnotation.ReadFieldsFirst("sourceId;targetId")
```

Comment:
Is it forwarding for source- and target-id, if a collector puts them into a DataSet and not writing them to another field of a Tuple for example?
I'm not quite sure here, too.
@galpha any suggestions?

Reply:
`ReadFields` is used to set all fields read during the computation of the result. Since you are checking the value of a certain property in an `if` statement, properties are also read. So you have to add properties to that annotation.
I think this should be
```java
@FunctionAnnotation.ReadFields("properties")
@FunctionAnnotation.ForwardedFields("sourceId;targetId")
```
although I'm unsure about forwarding here, since it does not always forward the field (but it never modifies it).

Also remember that those annotations are optional, remove them if you unsure.

Reply:
Ok - I'm going to just read the properties here.